js/json_ld_validator.js

```diff
@@ -0,0 +1,261 @@
+                    );
+                } // incorrect property name
+
+                if (!propDetails) {
+                    return removeProp(obj, key);
+                }
+
+                // datatype not followed
+                if (Array.isArray(obj[key])) {
+                    obj[key].forEach(function (element, index) {
+                        if (!isValidType(propDetails, element, key)) {
+                            module._log.warn("Deleting invalid element inside array for " + key + "\n");
+                            obj[key].splice(index, 1);
+                        }
+                    });
+                } else {
+                    if (!isValidType(propDetails, obj[key], key)) {
+                        removeProp(obj, key);
+                    }
+                }
+            }
+        });
+        return true;
+    }
+}
+
+/**
+ * 
+ * @param {*} obj The object from which the key/attribute should be removed
+ * @param {*} key The attribute name
+ */
+function removeProp(obj, key) {
+    module._log.info("Invalid attribute ignored " + key + " inside type " + obj[typeKeyword] + " in JSON-LD\n");
+    delete obj[key];
+}
+
+/**
+ * 
+ * @param {*} propDetails The acceptable data types for the element extracted from jsonLdSchema.js
+ * @param {*} element The attribute of the JSON-LD that we are currently validating for data type
+ * @param {*} key The attribute name
+ * @returns boolean that indicates if element follows its assigned type
+ */
+function isValidType(propDetails, element, key) {
+    var result = false;
+
+    if (propDetails !== null && propDetails !== undefined && propDetails.types) {
+        propDetails.types.every(function (dataType) {
+            switch (dataType) {
+                case "Text":
+                case "URL":
+                    if (typeof element == "string") {
+                        result = true;
+                        // returning false so we can exit out of every() loop
+                        return false;
+                    }
+                    break;
+
+                case "Date":
+                case "DateTime":
+                    if (Date.parse(element)) {
+                        result = true;
+                        return false;
+                    }
+                    break;
+
+                case "Number":
+                case "Integer":
+                    if (typeof element == "number") {
+                        result = true;
+                        return false;
+                    }
+
+                    break;
+
+                case "Boolean":
+                    if (typeof element == "boolean" || (typeof element == "string" && (element.toLowerCase() == "true" || element.toLowerCase() == "false"))) {
+                        result = true;
+                        return false;
+                    }
+                    break;
+            }
+
+            return true;
+        });
+
+        if (element[typeKeyword]) {
```

Comment:
Please add an inline comment explaining why you're doing this check and what this is doing